test/test_overrides.py

```diff
@@ -0,0 +1,814 @@
+def sub_diagonal_mm(mat1, mat2):
+    return 1
+
+@implements_sub_diagonal(torch.div)
+def sub_diagonal_div(input, other, out=None):
+    return NotImplemented
+
+# The dispatch table for SubDiagonalTensor's __torch_function__ implementation.
+HANDLED_FUNCTIONS_TENSOR_LIKE = {}
+
+def implements_tensor_like(torch_function):
+    "Register a torch function override for TensorLike"
+    @functools.wraps(torch_function)
+    def decorator(func):
+        HANDLED_FUNCTIONS_TENSOR_LIKE[torch_function] = func
+        return func
+    return decorator
+
+# Functions that are publicly available in the torch API but cannot be
+# overrided with __torch_function__ (usually because none of their
+# arguments are tensors or tensor-likes) need an entry in this tuple.
+
+IGNORED_TORCH_FUNCTIONS = (
+    torch.typename,
+    torch.is_tensor,
+    torch.is_storage,
+    torch.set_default_tensor_type,
+    torch.set_rng_state,
+    torch.get_rng_state,
+    torch.manual_seed,
+    torch.initial_seed,
+    torch.seed,
+    torch.save,
+    torch.load,
+    torch.set_printoptions,
+    torch.fork,
+    torch.get_default_dtype,
+    torch.get_num_interop_threads,
+    torch.get_num_threads,
+    torch.import_ir_module,
+    torch.import_ir_module_from_buffer,
+    torch.is_anomaly_enabled,
+    torch.is_grad_enabled,
+    torch.merge_type_from_type_comment,
+    torch.parse_ir,
+    torch.parse_schema,
+    torch.parse_type_comment,
+    torch.set_anomaly_enabled,
+    torch.set_flush_denormal,
+    torch.set_num_interop_threads,
+    torch.set_num_threads,
+    torch.wait,
+    torch.as_tensor,
+    torch.from_numpy,
+    torch.get_device,
+    torch.tensor,
+    torch.default_generator,
+    torch.has_cuda,
+    torch.has_cudnn,
+    torch.has_lapack,
+    torch.cpp,
+    torch.device,
+    torch.dtype,
+    torch.finfo,
+    torch.has_mkl,
+    torch.has_mkldnn,
+    torch.has_openmp,
+    torch.iinfo,
+    torch.memory_format,
+    torch.qscheme,
+    torch.set_grad_enabled,
+    torch.no_grad,
+    torch.enable_grad,
+    torch.layout,
+    torch.align_tensors,
+    torch.arange,
+    torch.as_strided,
+    torch.bartlett_window,
+    torch.blackman_window,
+    torch.can_cast,
+    torch.cudnn_affine_grid_generator,
+    torch.cudnn_batch_norm,
+    torch.cudnn_convolution,
+    torch.cudnn_convolution_transpose,
+    torch.cudnn_grid_sampler,
+    torch.cudnn_is_acceptable,
+    torch.empty,
+    torch.empty_strided,
+    torch.eye,
+    torch.from_file,
+    torch.full,
+    torch.hamming_window,
+    torch.hann_window,
+    torch.linspace,
+    torch.logspace,
+    torch.mkldnn_adaptive_avg_pool2d,
+    torch.mkldnn_convolution,
+    torch.mkldnn_convolution_backward_weights,
+    torch.mkldnn_max_pool2d,
+    torch.ones,
+    torch.promote_types,
+    torch.rand,
+    torch.randn,
+    torch.randint,
+    torch.randperm,
+    torch.range,
+    torch.zeros,
+)
+
+# Every function in the torch API that can be overriden needs an entry
+# in this tuple.
```

Comment:
Something people do on a fairly regular basis is add new functions to PyTorch. What happens to this test when that happens; does the test fail so that they have to go add their function to this list?

Reply:
Yup, here's the error message they'd see (here I've just removed the test for `torch.add`):

```
Traceback (most recent call last):
  File "test/test_overrides.py", line 704, in <module>
    generate_tensor_like_torch_implementations()
  File "test/test_overrides.py", line 699, in generate_tensor_like_torch_implementations
    assert len(untested_funcs) == 0, msg.format(pprint.pformat(untested_funcs))
AssertionError: The following functions are not tested for __torch_function__ support, please either add an entry in TENSOR_LIKE_TORCH_IMPLEMENTATIONS for this function or if a __torch_function__ override does not make sense, add an entry to IGNORED_TORCH_FUNCTIONS.

['torch.add']
```